core/src/main/scala/org/apache/spark/util/SizeEstimator.scala

```diff
@@ -32,6 +32,16 @@ import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.util.collection.OpenHashSet
 
 /**
+ * A trait that allows a class to give [[SizeEstimator]] more accurate size estimation.
+ * When a class extends it, [[SizeEstimator]] will query the `estimatedSize` first.
+ * If `estimatedSize` does not return [[None]], [[SizeEstimator]] will use the returned size
+ * as the size of the object. Otherwise, [[SizeEstimator]] will do the estimation work.
+ */
+private[spark] trait SizeEstimation {
+  def estimatedSize: Option[Long]
```

Comment:
I'd like to get it merged first if there is no fundamental issue. So, we can unblock the preview package. I can make the change if we prefer to change `BytesToBytesMap` instead of `UnsafeHashedRelation`. I agree returning a `Option` is weird. But, I feel if it is possible, we should prefer changing `UnsafeHashedRelation` because it is the one used as the broadcast variable.